IA-3211 Automation tests for createVM flow

Qi77Qi opened the pull request:
https://broadworkbench.atlassian.net/browse/IA-3211

Permission check for getRuntime is commented out for now because we'll need a change in WSM to make things work. Dan D is on vacation right now, so punt on that for this PR.

---
Have you read [CONTRIBUTING.md](https://github.com/DataBiosphere/leonardo/blob/develop/CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've documented my API changes in Swagger

In all cases:

- [ ] Get a thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green (you can re-run automation tests with a comment saying `jenkins retest`
- [ ] Run the automation tests multiple times in parallel to weed out instability if applicable via a comment saying `multi-test`
- [ ] Squash and merge; Delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

Commit:
initial commit, not compiling

Commit:
pre-vaca commit

Qi77Qi committed:
updates

Qi77Qi committed:
somewhat working

Qi77Qi committed:
working

Qi77Qi committed:
formatting

Qi77Qi committed:
cleanup

Qi77Qi committed:
fix test

Qi77Qi committed:
fix tests

Qi77Qi commented:
jenkins retest

Qi77Qi committed:
fix delete runtime

Qi77Qi committed:
delete associated resources

Qi77Qi committed:
fix tests

Qi77Qi committed:
fix tests

Qi77Qi commented:
jenkins retest

Qi77Qi commented:
fiab-start failure ... 
jenkins retest

Qi77Qi committed:
dummy to trigger test

Qi77Qi committed:
diable leo registration for now

Qi77Qi committed:
don't run azure test

jdcanas commented on automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoApiClient.scala:
```diff
@@ -116,6 +118,20 @@ object LeonardoApiClient {
+  val defaultCreateAzureRuntimeRequest = CreateAzureRuntimeRequest(
+    Map.empty,
+    Region.US_WEST_CENTRAL,
+    VirtualMachineSizeTypes.STANDARD_D1_V2, //Standard_D1_v2
```
can we remove the comment?

jdcanas commented on automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoApiClient.scala:
```diff
@@ -646,6 +662,90 @@ object LeonardoApiClient {
+        )
+        .use { resp =>
+          if (!resp.status.isSuccess) {
+            onError(s"Failed to create runtime ${workspaceId.value.toString}/${runtimeName.asString}")(resp)
```
error message is copy pasta

jdcanas commented on automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/azure/AzureRuntimeSpec.scala:
```diff
@@ -0,0 +1,59 @@
+    val res = LeonardoApiClient.client.use { implicit client =>
+      for {
+        _ <- loggerIO.info(s"AzureRuntimeSpec: About to create runtime")
+//        rat <- Ron.authToken()
```
can we remove this line?

Qi77Qi replied:
I left it there just becuz I think in the future when we need to create workspace, we'll need Ron's token

jdcanas commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpSamDAO.scala:
```diff
@@ -48,6 +46,29 @@ class HttpSamDAO[F[_]](httpClient: Client[F],
+          Request[F](
+            method = Method.POST,
+            uri = config.samUri.withPath(Uri.Path.unsafeFromString(s"/register/user/v2/self")),
+            body = Stream.emits("app.terra.bio/#terms-of-service".getBytes(UTF_8)),
```
We talked about why this call is necessary to register leo as an application in SAM, but what exactly does POSTing this particular payload do? Is it a string sam is expecting?

Qi77Qi replied:
it's just a required field for [this sam api](https://sam.dsde-dev.broadinstitute.org/#/Users/createUserV2)...it's always the same...not sure exactly what it does

jdcanas commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpWsmDao.scala:
```diff
@@ -96,38 +106,59 @@ class HttpWsmDao[F[_]](httpClient: Client[F], config: HttpWsmDaoConfig)(implicit
 
-  override def getCreateVmJobResult(request: GetJobResultRequest)(implicit ev: Ask[F, AppContext]): F[CreateVmResult] =
-    httpClient.expectOr[CreateVmResult](
+  private def deleteHelper(req: DeleteWsmResourceRequest, authorization: Authorization, resource: String)(
```
I'd consider using the controlled resource enum [here](https://github.com/DataBiosphere/leonardo/blob/develop/http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/RuntimeControlledResourceComponent.scala#L50) over a string (after possibly moving the enum to a more general location)

Qi77Qi replied:
the param is named as `resource`, but it's really just the url path for WSM api...I don't think it has much to do with `WsmResourceType`

jdcanas commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/SamDAO.scala:
```diff
@@ -71,6 +73,8 @@ trait SamDAO[F[_]] {
 
   def getUserSubjectIdFromToken(token: String)(implicit ev: Ask[F, TraceId]): F[Option[UserSubjectId]]
+
+  def getLeoAuthToken: F[Authorization]
```
hmm... it seems odd that this is an exposed trait of the SamDAO. I don't see it being used outside of HttpSamDao?

Qi77Qi replied:
it's used https://github.com/DataBiosphere/leonardo/pull/2581/files#diff-8af5da8d5edb93397a2e5c102c53047173dac803b46343c12568196a90cee553R211...I comment out that part for now becuz we in fiabs, sam is not started yet, so this will fail Leonardo's bootup.

jdcanas commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/Boot.scala:
```diff
@@ -209,6 +206,9 @@ object Boot extends IOApp {
         implicit0(ctx: Ask[IO, AppContext]) = Ask.const[IO, AppContext](
           AppContext(TraceId(s"Boot_${start}"), start)
         )
+//       TODO: this will be needed once we support more environments.
```
I feel like this TODO should have a ticket attached to it in the comment so we can ctrl+f for it when the time comes

Qi77Qi replied:
I agree in general. The real issue for this is in fiabs, Sam is not guranteed to boot up before Leonardo is. I'm not entirely sure how we get around that yet. So punting on creating a concrete ticket...but this TODO will definitely be addressed at some point in Q2, so I'm not too concerned about having a ticket since we'll have to address this as we roll out to more envs.

jdcanas commented on http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AzureServiceInterpSpec.scala:
```diff
@@ -203,14 +231,22 @@ class AzureServiceInterpSpec extends AnyFlatSpec with LeonardoTestSuite with Tes
         )
+        .transaction
+      cluster = clusterOpt.get
+      _ <- updateCreator(cluster.id, WorkbenchEmail("randomUser")).transaction
```
why do we update the creator here?

Qi77Qi replied:
just to fake it that the runtime is created by another user

Qi77Qi replied:
guess I can use a different user to call `getRuntime`...will do that

Qi77Qi committed:
comments

jdcanas reviewed (approved):
Overall looks good, a couple nits/questions

Qi77Qi commented:
thanks @jdcanas for reviews...think I addressed all of the comments that I plan to address...the rest I'm not entirely sure about...since you approved the PR already, I'm assuming you don't feel strongly about those. Going to merge the PR
